--- /Users/jhpark/VSLAM/ROVTIO_Docker/third_party/rovio/include/rovio/ImuPrediction.hpp	2026-08-18 02:04:42
+++ /Users/jhpark/VSLAM/ROVTIO_Docker/contrib/include/rovio/ImuPrediction.hpp	2026-08-18 03:08:23
@@ -50,13 +50,15 @@
   typedef typename Base::mtFilterState mtFilterState;
   typedef typename Base::mtMeas mtMeas;
   typedef typename Base::mtNoise mtNoise;
-  const V3D g_; /**<Gravity in inertial frame, always aligned with the z-axis.*/
+  V3D g_; /**<Gravity in inertial frame, always aligned with the z-axis.*/ // Custom: Not const anymore
   double inertialMotionRorTh_; /**<Threshold on the rotational rate for motion detection.*/
   double inertialMotionAccTh_; /**<Threshold on the acceleration for motion detection.*/
   mutable FeatureCoordinates oldC_;
   mutable FeatureDistance oldD_;
   mutable Eigen::Matrix2d bearingVectorJac_;
+//  ImuPrediction():g_(0,0,-9.81){ // Original
   ImuPrediction():g_(0,0,-9.81){
+    doubleRegister_.registerVector("GravityVector.g", g_); // Custom to have custom gravity vectors
     int ind;
     inertialMotionRorTh_ = 0.1;
     inertialMotionAccTh_ = 0.1;
