--- /Users/jhpark/VSLAM/ROVTIO_Docker/third_party/rovio/include/rovio/MultilevelPatchAlignment.hpp	2026-08-18 02:04:42
+++ /Users/jhpark/VSLAM/ROVTIO_Docker/contrib/include/rovio/MultilevelPatchAlignment.hpp	2026-08-18 03:08:23
@@ -360,145 +360,145 @@
    * @return true, if alignment converged!
    * @todo catch if warping too distorted
    */
-  bool align2D_old(FeatureCoordinates& cOut, const ImagePyramid<nLevels>& pyr, const MultilevelPatch<nLevels,patch_size>& mp, const FeatureCoordinates& cInit, const int l1, const int l2,
-                   const int maxIter = 10, const double minPixUpd = 0.03){
-    Eigen::Matrix2f WInv;
-    if(!cInit.com_c() || !cInit.com_warp_c()){
-      return false;
-    }
-    WInv = cInit.get_warp_c().inverse();
-    int numLevel = 0;
-    FeatureCoordinates c_level;
-    for(int l = l1; l <= l2; l++){
-      pyr.levelTranformCoordinates(cInit,c_level,0,l);
-      if(mp.isValidPatch_[l] && mp.patches_[l].isPatchInFrame(pyr.imgs_[l],c_level,false)){
-        mp.patches_[l].computeGradientParameters();
-        if(mp.patches_[l].validGradientParameters_){
-          numLevel++;
-        }
-      }
-    }
-    if(numLevel==0){
-      return false;
-    }
-    Eigen::Matrix3f aff = Eigen::Matrix3f::Identity();
-    aff.block<2,2>(0,0) = cInit.get_warp_c();
-    Eigen::Matrix3f affInv = aff.inverse();
-    const int halfpatch_size = patch_size/2;
-    bool converged=false;
-    Eigen::Matrix3f H; H.setZero();
-    for(int l = l1; l <= l2; l++){
-      pyr.levelTranformCoordinates(cInit,c_level,0,l);
-      if(mp.isValidPatch_[l] && mp.patches_[l].isPatchInFrame(pyr.imgs_[l],c_level,false)){
-        mp.patches_[l].computeGradientParameters();
-        H(0,0) += pow(0.25,l)*mp.patches_[l].H_(0,0);
-        H(0,1) += pow(0.25,l)*mp.patches_[l].H_(0,1);
-        H(1,0) += pow(0.25,l)*mp.patches_[l].H_(1,0);
-        H(1,1) += pow(0.25,l)*mp.patches_[l].H_(1,1);
-        H(2,0) += pow(0.5,l)*mp.patches_[l].H_(2,0);
-        H(2,1) += pow(0.5,l)*mp.patches_[l].H_(2,1);
-        H(0,2) += pow(0.5,l)*mp.patches_[l].H_(0,2);
-        H(1,2) += pow(0.5,l)*mp.patches_[l].H_(1,2);
-        H(2,2) += mp.patches_[l].H_(2,2);
-      }
-    }
-    Eigen::Matrix3f Hinv = H.inverse();
-    float mean_diff = 0;
-
-    // termination condition
-    const float min_update_squared = minPixUpd*minPixUpd;
-    cOut = cInit;
-    Eigen::Vector3f update; update.setZero();
-    for(int iter = 0; iter<maxIter; ++iter){
-      if(std::isnan(cOut.get_c().x) || std::isnan(cOut.get_c().y)){
-        assert(false);
-        return false;
-      }
-      Eigen::Vector3f Jres; Jres.setZero();
-      int count = 0;
-      for(int l = l1; l <= l2; l++){
-        pyr.levelTranformCoordinates(cOut,c_level,0,l);
-        if(mp.isValidPatch_[l] && mp.patches_[l].isPatchInFrame(pyr.imgs_[l],c_level,false)){
-          const int refStep = pyr.imgs_[l].step.p[0];
-
-          const float* it_patch = mp.patches_[l].patch_;
-          const float* it_dx = mp.patches_[l].dx_;
-          const float* it_dy = mp.patches_[l].dy_;
-          if(cInit.isNearIdentityWarping()){
-            const int u_r = floor(c_level.get_c().x);
-            const int v_r = floor(c_level.get_c().y);
-            if(u_r < halfpatch_size || v_r < halfpatch_size || u_r >= pyr.imgs_[l].cols-halfpatch_size || v_r >= pyr.imgs_[l].rows-halfpatch_size){ // TODO: check limits
-              return false;
-            }
-            // compute interpolation weights
-            const float subpix_x = c_level.get_c().x-u_r;
-            const float subpix_y = c_level.get_c().y-v_r;
-            const float wTL = (1.0-subpix_x)*(1.0-subpix_y);
-            const float wTR = subpix_x * (1.0-subpix_y);
-            const float wBL = (1.0-subpix_x)*subpix_y;
-            const float wBR = subpix_x * subpix_y;
-
-            const uint8_t* it_img;        // loop through search_patch, interpolate
-            for(int y=0; y<patch_size; ++y){
-              it_img = (uint8_t*) pyr.imgs_[l].data + (v_r+y-halfpatch_size)*refStep + u_r-halfpatch_size;
-              for(int x=0; x<patch_size; ++x, ++it_img, ++it_patch, ++it_dx, ++it_dy){
-                const float intensity = wTL*it_img[0] + wTR*it_img[1] + wBL*it_img[refStep] + wBR*it_img[refStep+1];
-                const float res = intensity - *it_patch + mean_diff;
-                Jres[0] -= pow(0.5,l)*res*(*it_dx);
-                Jres[1] -= pow(0.5,l)*res*(*it_dy);
-                Jres[2] -= res;
-              }
-            }
-          } else {
-            for(int y=0; y<patch_size; ++y){
-              for(int x=0; x<patch_size; ++x, ++it_patch, ++it_dx, ++it_dy){
-                const float dx = x - halfpatch_size + 0.5;
-                const float dy = y - halfpatch_size + 0.5;
-                const float wdx = cInit.get_warp_c()(0,0)*dx + cInit.get_warp_c()(0,1)*dy;
-                const float wdy = cInit.get_warp_c()(1,0)*dx + cInit.get_warp_c()(1,1)*dy;
-                const float u_pixel = c_level.get_c().x + wdx - 0.5;
-                const float v_pixel = c_level.get_c().y + wdy - 0.5;
-                const int u_pixel_r = floor(u_pixel);
-                const int v_pixel_r = floor(v_pixel);
-                if(u_pixel_r < 0 || v_pixel_r < 0 || u_pixel_r >= pyr.imgs_[l].cols-1 || v_pixel_r >= pyr.imgs_[l].rows-1){ // TODO: check limits
-                  return false;
-                }
-                const float pixel_subpix_x = u_pixel-u_pixel_r;
-                const float pixel_subpix_y = v_pixel-v_pixel_r;
-                const float pixel_wTL = (1.0-pixel_subpix_x) * (1.0-pixel_subpix_y);
-                const float pixel_wTR = pixel_subpix_x * (1.0-pixel_subpix_y);
-                const float pixel_wBL = (1.0-pixel_subpix_x) * pixel_subpix_y;
-                const float pixel_wBR = pixel_subpix_x * pixel_subpix_y;
-                const uint8_t* pixel_data = (uint8_t*) pyr.imgs_[l].data + v_pixel_r*refStep + u_pixel_r;
-                const float pixel_intensity = pixel_wTL*pixel_data[0] + pixel_wTR*pixel_data[1] + pixel_wBL*pixel_data[refStep] + pixel_wBR*pixel_data[refStep+1];
-                const float res = pixel_intensity - *it_patch + mean_diff;
-                Jres[0] -= pow(0.5,l)*res*(*it_dx);
-                Jres[1] -= pow(0.5,l)*res*(*it_dy);
-                Jres[2] -= res;
-              }
-            }
-          }
-          count++;
-        }
-      }
-      if(count==0){
-        return false;
-      }
-      if(cInit.isNearIdentityWarping()){
-        update = Hinv * Jres;
-      } else {
-        update = aff * Hinv * Jres;
-      }
-      cOut.set_c(cv::Point2f(cOut.get_c().x + update[0],cOut.get_c().y + update[1]),false);
-      mean_diff += update[2];
-
-      if(update[0]*update[0]+update[1]*update[1] < min_update_squared){
-        converged=true;
-        break;
-      }
-    }
-    return converged;
-  }
+//  bool align2D_old(FeatureCoordinates& cOut, const ImagePyramid<nLevels>& pyr, const MultilevelPatch<nLevels,patch_size>& mp, const FeatureCoordinates& cInit, const int l1, const int l2,
+//                   const int maxIter = 10, const double minPixUpd = 0.03){
+//    Eigen::Matrix2f WInv;
+//    if(!cInit.com_c() || !cInit.com_warp_c()){
+//      return false;
+//    }
+//    WInv = cInit.get_warp_c().inverse();
+//    int numLevel = 0;
+//    FeatureCoordinates c_level;
+//    for(int l = l1; l <= l2; l++){
+//      pyr.levelTranformCoordinates(cInit,c_level,0,l);
+//      if(mp.isValidPatch_[l] && mp.patches_[l].isPatchInFrame(pyr.imgs_[l],c_level,false)){
+//        mp.patches_[l].computeGradientParameters();
+//        if(mp.patches_[l].validGradientParameters_){
+//          numLevel++;
+//        }
+//      }
+//    }
+//    if(numLevel==0){
+//      return false;
+//    }
+//    Eigen::Matrix3f aff = Eigen::Matrix3f::Identity();
+//    aff.block<2,2>(0,0) = cInit.get_warp_c();
+//    Eigen::Matrix3f affInv = aff.inverse();
+//    const int halfpatch_size = patch_size/2;
+//    bool converged=false;
+//    Eigen::Matrix3f H; H.setZero();
+//    for(int l = l1; l <= l2; l++){
+//      pyr.levelTranformCoordinates(cInit,c_level,0,l);
+//      if(mp.isValidPatch_[l] && mp.patches_[l].isPatchInFrame(pyr.imgs_[l],c_level,false)){
+//        mp.patches_[l].computeGradientParameters();
+//        H(0,0) += pow(0.25,l)*mp.patches_[l].H_(0,0);
+//        H(0,1) += pow(0.25,l)*mp.patches_[l].H_(0,1);
+//        H(1,0) += pow(0.25,l)*mp.patches_[l].H_(1,0);
+//        H(1,1) += pow(0.25,l)*mp.patches_[l].H_(1,1);
+//        H(2,0) += pow(0.5,l)*mp.patches_[l].H_(2,0);
+//        H(2,1) += pow(0.5,l)*mp.patches_[l].H_(2,1);
+//        H(0,2) += pow(0.5,l)*mp.patches_[l].H_(0,2);
+//        H(1,2) += pow(0.5,l)*mp.patches_[l].H_(1,2);
+//        H(2,2) += mp.patches_[l].H_(2,2);
+//      }
+//    }
+//    Eigen::Matrix3f Hinv = H.inverse();
+//    float mean_diff = 0;
+//
+//    // termination condition
+//    const float min_update_squared = minPixUpd*minPixUpd;
+//    cOut = cInit;
+//    Eigen::Vector3f update; update.setZero();
+//    for(int iter = 0; iter<maxIter; ++iter){
+//      if(std::isnan(cOut.get_c().x) || std::isnan(cOut.get_c().y)){
+//        assert(false);
+//        return false;
+//      }
+//      Eigen::Vector3f Jres; Jres.setZero();
+//      int count = 0;
+//      for(int l = l1; l <= l2; l++){
+//        pyr.levelTranformCoordinates(cOut,c_level,0,l);
+//        if(mp.isValidPatch_[l] && mp.patches_[l].isPatchInFrame(pyr.imgs_[l],c_level,false)){
+//          const int refStep = pyr.imgs_[l].step.p[0];
+//
+//          const float* it_patch = mp.patches_[l].patch_;
+//          const float* it_dx = mp.patches_[l].dx_;
+//          const float* it_dy = mp.patches_[l].dy_;
+//          if(cInit.isNearIdentityWarping()){
+//            const int u_r = floor(c_level.get_c().x);
+//            const int v_r = floor(c_level.get_c().y);
+//            if(u_r < halfpatch_size || v_r < halfpatch_size || u_r >= pyr.imgs_[l].cols-halfpatch_size || v_r >= pyr.imgs_[l].rows-halfpatch_size){ // TODO: check limits
+//              return false;
+//            }
+//            // compute interpolation weights
+//            const float subpix_x = c_level.get_c().x-u_r;
+//            const float subpix_y = c_level.get_c().y-v_r;
+//            const float wTL = (1.0-subpix_x)*(1.0-subpix_y);
+//            const float wTR = subpix_x * (1.0-subpix_y);
+//            const float wBL = (1.0-subpix_x)*subpix_y;
+//            const float wBR = subpix_x * subpix_y;
+//
+//            const uint8_t* it_img;        // loop through search_patch, interpolate
+//            for(int y=0; y<patch_size; ++y){
+//              it_img = (uint8_t*) pyr.imgs_[l].data + (v_r+y-halfpatch_size)*refStep + u_r-halfpatch_size;
+//              for(int x=0; x<patch_size; ++x, ++it_img, ++it_patch, ++it_dx, ++it_dy){
+//                const float intensity = wTL*it_img[0] + wTR*it_img[1] + wBL*it_img[refStep] + wBR*it_img[refStep+1];
+//                const float res = intensity - *it_patch + mean_diff;
+//                Jres[0] -= pow(0.5,l)*res*(*it_dx);
+//                Jres[1] -= pow(0.5,l)*res*(*it_dy);
+//                Jres[2] -= res;
+//              }
+//            }
+//          } else {
+//            for(int y=0; y<patch_size; ++y){
+//              for(int x=0; x<patch_size; ++x, ++it_patch, ++it_dx, ++it_dy){
+//                const float dx = x - halfpatch_size + 0.5;
+//                const float dy = y - halfpatch_size + 0.5;
+//                const float wdx = cInit.get_warp_c()(0,0)*dx + cInit.get_warp_c()(0,1)*dy;
+//                const float wdy = cInit.get_warp_c()(1,0)*dx + cInit.get_warp_c()(1,1)*dy;
+//                const float u_pixel = c_level.get_c().x + wdx - 0.5;
+//                const float v_pixel = c_level.get_c().y + wdy - 0.5;
+//                const int u_pixel_r = floor(u_pixel);
+//                const int v_pixel_r = floor(v_pixel);
+//                if(u_pixel_r < 0 || v_pixel_r < 0 || u_pixel_r >= pyr.imgs_[l].cols-1 || v_pixel_r >= pyr.imgs_[l].rows-1){ // TODO: check limits
+//                  return false;
+//                }
+//                const float pixel_subpix_x = u_pixel-u_pixel_r;
+//                const float pixel_subpix_y = v_pixel-v_pixel_r;
+//                const float pixel_wTL = (1.0-pixel_subpix_x) * (1.0-pixel_subpix_y);
+//                const float pixel_wTR = pixel_subpix_x * (1.0-pixel_subpix_y);
+//                const float pixel_wBL = (1.0-pixel_subpix_x) * pixel_subpix_y;
+//                const float pixel_wBR = pixel_subpix_x * pixel_subpix_y;
+//                const uint8_t* pixel_data = (uint8_t*) pyr.imgs_[l].data + v_pixel_r*refStep + u_pixel_r;
+//                const float pixel_intensity = pixel_wTL*pixel_data[0] + pixel_wTR*pixel_data[1] + pixel_wBL*pixel_data[refStep] + pixel_wBR*pixel_data[refStep+1];
+//                const float res = pixel_intensity - *it_patch + mean_diff;
+//                Jres[0] -= pow(0.5,l)*res*(*it_dx);
+//                Jres[1] -= pow(0.5,l)*res*(*it_dy);
+//                Jres[2] -= res;
+//              }
+//            }
+//          }
+//          count++;
+//        }
+//      }
+//      if(count==0){
+//        return false;
+//      }
+//      if(cInit.isNearIdentityWarping()){
+//        update = Hinv * Jres;
+//      } else {
+//        update = aff * Hinv * Jres;
+//      }
+//      cOut.set_c(cv::Point2f(cOut.get_c().x + update[0],cOut.get_c().y + update[1]),false);
+//      mean_diff += update[2];
+//
+//      if(update[0]*update[0]+update[1]*update[1] < min_update_squared){
+//        converged=true;
+//        break;
+//      }
+//    }
+//    return converged;
+//  }
 
   /** \brief 2D patch alignment. No guarantee that final coordinates are fully in the frame.
    *
